### PR TITLE
VirtualNode: deletion-routine refactoring

### PR DESCRIPTION
**Pull request by cheina97**
# Description

This PR refactors the deletion-routine based on kubernetes guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md

### CODE REVIEW

**Comment by cheina97** [comment]
https://github.com/cheina97/liqo/blob/6959bd9c73fcc5765625f1f52b7766e3a6f689a1/pkg/liqo-controller-manager/virtualnode-controller/virtualkubelet.go#L167-L221

I'm not sure this is a good place for these functions. Do you think pkg/utils/virtualnode or pkg/vkMachinery could be better places?

File: pkg/liqo-controller-manager/virtualnode-controller/virtualkubelet.go [hunk 1]
```diff
@@ -177,15 +186,36 @@ func (r *VirtualNodeReconciler) getVirtualKubeletDeployment(
...
+	list, err := getters.ListVirtualKubeletPodsFromVirtualNode(ctx, r.Client, vn, r.VirtualKubeletOptions)
 	if err != nil {
-		klog.Error(err)
-		return false, nil
+		return false, err
 	}
 	return len(list.Items) == 0, err
 }
+
+func (r *VirtualNodeReconciler) checkVirtualKubeletPodsContainsOnlyArg(
```
Comment (aleoli) [inline pkg/liqo-controller-manager/virtualnode-controller/virtualkubelet.go:199]: Can you choose a better function naming and/or add some comments on what it is doing?

File: pkg/liqo-controller-manager/virtualnode-controller/deletion-routine.go [hunk 2]
```diff
@@ -53,64 +57,114 @@ func RunDeletionRoutine(ctx context.Context, r *VirtualNodeReconciler) (*Deletio
...
+
+func (dr *DeletionRoutine) handle(ctx context.Context, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("error splitting key: %w", err)
+	}
+	ref := types.NamespacedName{Namespace: namespace, Name: name}
+	vn := &virtualkubeletv1alpha1.VirtualNode{}
+	if err := dr.vnr.Client.Get(ctx, ref, vn); err != nil {
+		if client.IgnoreNotFound(err) == nil {
```
Comment (aleoli) [inline pkg/liqo-controller-manager/virtualnode-controller/deletion-routine.go:108]:
```suggestion
		if k8serrors.IsNotFound(err) {
```

File: pkg/liqo-controller-manager/virtualnode-controller/deletion-routine.go [hunk 3]
```diff
@@ -53,64 +57,114 @@ func RunDeletionRoutine(ctx context.Context, r *VirtualNodeReconciler) (*Deletio
...
-				dr.reEnqueueVirtualNode(vn, fmt.Errorf("error removing finalizer: %w", err))
-				continue
+	node, err := getters.GetNodeFromVirtualNode(ctx, dr.vnr.Client, vn)
+	if client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("error getting node: %w", err)
+	}
+
+	if node != nil {
+		if !*vn.Spec.CreateNode {
+			// We need to check if the schedule pods could recreate the node before deleting it.
```
Comment (aleoli) [inline pkg/liqo-controller-manager/virtualnode-controller/deletion-routine.go:130]:
```suggestion
			// We need to ensure that the current pods will no recreate the node before deleting it.
```

File: pkg/liqo-controller-manager/virtualnode-controller/deletion-routine.go [hunk 4]
```diff
@@ -53,64 +57,114 @@ func RunDeletionRoutine(ctx context.Context, r *VirtualNodeReconciler) (*Deletio
...
+			if found, err := dr.vnr.checkVirtualKubeletPodsContainsOnlyArg(
+				ctx, vn, string(vkMachineryForge.CreateNode), strconv.FormatBool(*vn.Spec.CreateNode)); err != nil || !found {
+				if err == nil {
+					err = fmt.Errorf("virtual kubelet pods do not contain only arg %s=%s", vkMachineryForge.CreateNode, strconv.FormatBool(*vn.Spec.CreateNode))
```
Comment (aleoli) [inline pkg/liqo-controller-manager/virtualnode-controller/deletion-routine.go:134]: Can you improve the error message? I don't understand what it means

**Review by aleoli — APPROVED** [review]

File: pkg/liqo-controller-manager/virtualnode-controller/deletion-routine.go [hunk 5]
```diff
@@ -53,64 +57,114 @@ func RunDeletionRoutine(ctx context.Context, r *VirtualNodeReconciler) (*Deletio
...
-		if err := UpdateCondition(ctx, dr.vnr.Client, vn,
-			VkConditionMap{
-				virtualkubeletv1alpha1.NodeConditionType: VkCondition{
-					Status: virtualkubeletv1alpha1.DrainingConditionStatusType,
-				},
-			}); err != nil {
-			dr.reEnqueueVirtualNode(vn, fmt.Errorf("error updating condition: %w", err))
-			continue
-		}
+	defer klog.Fatal("Deletion routine stopped")
```
Comment (aleoli) [inline pkg/liqo-controller-manager/virtualnode-controller/deletion-routine.go:71]: Avoid using fatal, it will make the pod crash

**Pull request merged**